### DISCUSSION TITLE
CUDA out of memory when saving checkpoint

### DISCUSSION THREAD

**Discussion by TonJ24** [open-mmlab/mmediting#1199] [How-to] [answered]
I meet a issue that happened when saving the checkpoint.
Specifically, the cuda memory would keep increasing when saving the checkpoint.
![image](https://user-images.githubusercontent.com/71822595/171998435-87678b88-07c4-492c-a03d-9956f3f93249.png)
I wanna to ask for help about how to resolve it.
Thanks very much.

**Comment by wangruohui** [comment]
From your screenshoot, the program seems running a validation step. The checkpoints should has been saved. You can take a look at the working directory to see if some `.pth` file exists. 

Do you mean the memory keep increasing during the progress propagates? 
What metric do you use for validation?
Are you running a standard configuration file?

**Comment by TonJ24** [comment]
Thanks for your reply.
The .pth exists in my file.
I just run the source code from author, thus the metric is PSNR/SSIM.




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "open-mmlab/mmediting"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年6月4日(星期六) 晚上9:16
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [open-mmlab/mmediting] CUDA out of memory when saving checkpoint (Issue #911)





 
From your screenshoot, the program seems running a validation step. The checkpoints should has been saved. You can take a look at the working directory to see if some .pth file exists.
 
What metric do you use for validation?

**Comment by wangruohui** [comment]
realbasicvsr_c64b20_1x30x8_lr5e-5_150k_reds.py
or
realbasicvsr_wogan_c64b20_2x30x8_lr1e-4_300k_reds.py

We can try reproduce your problem

**Comment by wangruohui** [comment]
BTW, if the memory bumps when validation starts, and keep nearly a constant, it should be normal because in current implementation, validation is a standalone function launched apart from training. So it will take some memory. 
If the memory keep increasing as the progress bar progresses, there might be some memory leak, e.g. some result tensor not detached from the computation graph and it keeps whole computation graph, which takes memory.

**Comment by TonJ24** [comment]
Thanks for your reply
"validation is a standalone function launched apart from training. So it will take some memory."—— The increased memory produced by validation has been existing later, I wanna to know whether it is normal?




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "open-mmlab/mmediting"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年6月4日(星期六) 晚上9:27
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [open-mmlab/mmediting] CUDA out of memory when saving checkpoint (Issue #911)





 
BTW, if the memory bumps when validation starts, and keep nearly a constant, it should be normal because in current implementation, validation is a standalone function launched apart from training. So it will take some memory.
 If the memory keep increasing as the progress bar progresses, there might be some memory leak, e.g. some result tensor not detached from the computation graph and it keeps whole computation graph, which takes memory.

**Comment by TonJ24** [comment]
> 

I wanna to ask for help about how to modify current implementation for validation without taking some memory.

**Comment by wangruohui** [comment, marked as answer]
Sorry for some late as we are working on some heavy development. 

As a very simple workaround, you can just disable evaluation by setting `interval` of the evaluation larger than the total iter of training. 

```
evaluation = dict(interval=5000, save_image=False, gpu_collect=True)
```

As checkpoint saving works anyway, you can make some analysis after the training or in an offline way by using another GPU.
